src/call_context.rs

```diff
@@ -75,172 +72,174 @@ impl StackFrame {
     }
 }
 
-pub trait Invoke<H: ByteHash>: Sized {
+pub trait Invoke<S: Store>: Sized {
     fn invoke(
-        context: &mut CallContext<Self, H>,
+        context: &mut CallContext<Self, S>,
         index: usize,
         args: RuntimeArgs,
-    ) -> Result<Option<RuntimeValue>, VMError>;
+    ) -> Result<Option<RuntimeValue>, VMError<S>>;
 }
 
-pub struct CallContext<'a, S, H: ByteHash> {
-    state: &'a mut NetworkState<S, H>,
+pub struct CallContext<'a, E, S: Store> {
+    state: &'a mut NetworkState<E, S>,
     stack: Vec<StackFrame>,
+    store: S,
     gas_meter: &'a mut GasMeter,
-    top_argument: &'a [u8],
-    top_return: &'a mut [u8],
-    _marker: PhantomData<S>,
 }
 
-impl<'a, S, H> CallContext<'a, S, H>
+impl<'a, E, S> CallContext<'a, E, S>
 where
-    S: Resolver<H>,
-    H: ByteHash,
+    E: Resolver<S>,
+    S: Store,
 {
-    pub fn new<A: Pod, R: Pod>(
-        state: &'a mut NetworkState<S, H>,
+    pub fn new(
+        state: &'a mut NetworkState<E, S>,
         gas_meter: &'a mut GasMeter,
-        top_argument: &'a A,
-        top_return: &'a mut R,
-    ) -> Self {
-        CallContext {
+        store: &S,
+    ) -> Result<Self, VMError<S>> {
+        Ok(CallContext {
             state,
             stack: vec![],
             gas_meter,
-            top_argument: top_argument.as_bytes(),
-            top_return: top_return.as_bytes_mut(),
-            _marker: PhantomData,
-        }
+            store: store.clone(),
+        })
     }
 
-    pub fn call(
+    pub fn query(
         &mut self,
-        target: H256,
-        opcode: usize,
-        argument_ptr: usize,
-        argument_len: usize,
-        return_ptr: usize,
-        return_len: usize,
-    ) -> Result<Option<RuntimeValue>, VMError> {
-        let resolver = S::default();
+        target: &ContractId,
+        query: Query,
+    ) -> Result<ReturnValue, VMError<S>> {
+        let resolver = E::default();
         let imports = ImportsBuilder::new().with_resolver("env", &resolver);
 
         let instance;
 
-        match self.state.get_contract_state_mut(&target)? {
-            None => return Err(VMError::UnknownContract),
-            Some(mut contract_state) => {
-                contract_state.contract_mut().ensure_compiled()?;
+        let store = self.store.clone();
 
-                let module = match contract_state.contract() {
-                    MeteredContract::Module { ref module, .. } => module,
-                    _ => unreachable!(),
-                };
+        match self.state.get_contract(target)? {
+            None => panic!("FIXME: error handling"),
```

Comment:
The function returns a result but we opted to panic. Why not return `Err` instead?

Reply:
There's several places in the code that already panics where it shouldn't. I'll make a separate issue for this.